File: ios/App/App/NativeOpenAI/NativeOpenAIPlugin.swift
import Foundation
import UIKit
import Capacitor

/// Capacitor-Plugin "NativeOpenAI".
///
/// Sicherheitsgrundsätze:
/// - Der API-Key wird über einen nativen Dialog eingegeben und direkt in
///   der Keychain gespeichert – er passiert die JS-Bridge nicht.
/// - Es existiert bewusst keine getApiKey-Methode.
/// - Host, Pfad und Authorization-Header liegen nur im nativen Code.
@objc(NativeOpenAIPlugin)
public class NativeOpenAIPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeOpenAIPlugin"
    public let jsName = "NativeOpenAI"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "hasApiKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "promptForApiKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteApiKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testConnection", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createResponse", returnType: CAPPluginReturnPromise)
    ]

    /// Muss mit DEFAULT_AI_MODEL in src/ai/prompt-builder.ts übereinstimmen;
    /// wird nur für den Verbindungstest verwendet.
    private static let defaultModel = "gpt-5.6-terra"

    private let client = OpenAIClient()

    @objc func hasApiKey(_ call: CAPPluginCall) {
        call.resolve(["configured": KeychainStore.exists()])
    }

    @objc func promptForApiKey(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let viewController = self.bridge?.viewController else {
                call.reject("UNKNOWN", "UNKNOWN")
                return
            }

            let alert = UIAlertController(
                title: "OpenAI-Zugang",
                message: "Der Schlüssel wird nur in der Keychain dieses Geräts gespeichert.",
                preferredStyle: .alert
            )
            alert.addTextField { textField in
                textField.placeholder = "API-Key"
                textField.isSecureTextEntry = true
                textField.autocorrectionType = .no
                textField.autocapitalizationType = .none
            }
            alert.addAction(UIAlertAction(title: "Abbrechen", style: .cancel) { _ in
                call.resolve(["saved": false, "validated": false])
            })
            alert.addAction(UIAlertAction(title: "Speichern", style: .default) { [weak alert] _ in
                let entered = alert?.textFields?.first?.text?
                    .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                guard !entered.isEmpty else {
                    call.resolve(["saved": false, "validated": false])
                    return
                }
                guard KeychainStore.save(entered) else {
                    call.reject("UNKNOWN", "UNKNOWN")
                    return
                }
                // Der Key bleibt nativ; nur das Testergebnis geht zurück.
                self.client.testConnection(model: Self.defaultModel) { valid, _ in
                    call.resolve(["saved": true, "validated": valid])
                }
            })

            viewController.present(alert, animated: true)
        }
    }

    @objc func deleteApiKey(_ call: CAPPluginCall) {
        KeychainStore.delete()
        call.resolve()
    }

    @objc func testConnection(_ call: CAPPluginCall) {
        guard KeychainStore.exists() else {
            call.reject("API_KEY_MISSING", "API_KEY_MISSING")
            return
        }
        client.testConnection(model: Self.defaultModel) { valid, modelAccess in
            call.resolve(["valid": valid, "modelAccess": modelAccess])
        }
    }

    @objc func createResponse(_ call: CAPPluginCall) {
        guard let body = call.getString("body"), !body.isEmpty else {
            call.reject("OPENAI_BAD_REQUEST", "OPENAI_BAD_REQUEST")
            return
        }

        client.createResponse(body: body) { result in
            switch result {
            case .success(let response):
                call.resolve([
                    "status": response.status,
                    "body": response.body,
                    "requestId": response.requestId as Any
                ])
            case .failure(let error):
                call.reject(error.code, error.code)
            }
        }
    }
}

File: android/app/src/main/java/de/eigenraum/app/NativeOpenAIPlugin.java
package de.eigenraum.app;

import android.app.AlertDialog;
import android.text.InputType;
import android.widget.EditText;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

/**
 * Capacitor-Plugin "NativeOpenAI" (Android).
 *
 * Der API-Key wird in einem nativen Dialog eingegeben, AES-GCM-
 * verschlüsselt gespeichert und nie an JavaScript zurückgegeben.
 * Es existiert bewusst keine getApiKey-Methode.
 */
@CapacitorPlugin(name = "NativeOpenAI")
public class NativeOpenAIPlugin extends Plugin {

    /** Muss mit DEFAULT_AI_MODEL in src/ai/prompt-builder.ts übereinstimmen. */
    private static final String DEFAULT_MODEL = "gpt-5.6-terra";

    private KeyStorage keyStorage;
    private OpenAIClient client;

    @Override
    public void load() {
        keyStorage = new KeyStorage(getContext());
        client = new OpenAIClient();
    }

    @PluginMethod
    public void hasApiKey(PluginCall call) {
        JSObject result = new JSObject();
        result.put("configured", keyStorage.exists());
        call.resolve(result);
    }

    @PluginMethod
    public void promptForApiKey(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            EditText input = new EditText(getActivity());
            input.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
            input.setHint("API-Key");

            new AlertDialog.Builder(getActivity())
                    .setTitle("OpenAI-Zugang")
                    .setMessage("Der Schlüssel wird verschlüsselt auf diesem Gerät gespeichert.")
                    .setView(input)
                    .setNegativeButton("Abbrechen", (dialog, which) -> {
                        JSObject result = new JSObject();
                        result.put("saved", false);
                        result.put("validated", false);
                        call.resolve(result);
                    })
                    .setPositiveButton("Speichern", (dialog, which) -> {
                        String entered = input.getText().toString().trim();
                        if (entered.isEmpty()) {
                            JSObject result = new JSObject();
                            result.put("saved", false);
                            result.put("validated", false);
                            call.resolve(result);
                            return;
                        }
                        if (!keyStorage.save(entered)) {
                            call.reject("UNKNOWN", "UNKNOWN");
                            return;
                        }
                        testStoredKey((valid, modelAccess) -> {
                            JSObject result = new JSObject();
                            result.put("saved", true);
                            result.put("validated", valid);
                            call.resolve(result);
                        });
                    })
                    .setCancelable(false)
                    .show();
        });
    }

    @PluginMethod
    public void deleteApiKey(PluginCall call) {
        keyStorage.delete();
        call.resolve();
    }

    @PluginMethod
    public void testConnection(PluginCall call) {
        if (!keyStorage.exists()) {
            call.reject("API_KEY_MISSING", "API_KEY_MISSING");
            return;
        }
        testStoredKey((valid, modelAccess) -> {
            JSObject result = new JSObject();
            result.put("valid", valid);
            result.put("modelAccess", modelAccess);
            call.resolve(result);
        });
    }

    @PluginMethod
    public void createResponse(PluginCall call) {
        String body = call.getString("body");
        if (body == null || body.isEmpty()) {
            call.reject("OPENAI_BAD_REQUEST", "OPENAI_BAD_REQUEST");
            return;
        }
        String apiKey = keyStorage.read();
        if (apiKey == null || apiKey.isEmpty()) {
            call.reject("API_KEY_MISSING", "API_KEY_MISSING");
            return;
        }

        client.createResponse(apiKey, body, new OpenAIClient.Callback() {
            @Override
            public void onSuccess(int status, String responseBody, String requestId) {
                JSObject result = new JSObject();
                result.put("status", status);
                result.put("body", responseBody);
                result.put("requestId", requestId);
                call.resolve(result);
            }

            @Override
            public void onError(String code) {
                call.reject(code, code);
            }
        });
    }

    private interface TestResult {
        void done(boolean valid, boolean modelAccess);
    }

    /** Minimaler Verbindungstest über den fest codierten Endpoint. */
    private void testStoredKey(TestResult onDone) {
        String apiKey = keyStorage.read();
        if (apiKey == null || apiKey.isEmpty()) {
            onDone.done(false, false);
            return;
        }
        try {
            JSONObject probe = new JSONObject();
            probe.put("model", DEFAULT_MODEL);
            probe.put("input", "ping");
            probe.put("store", false);
            probe.put("max_output_tokens", 16);

            client.createResponse(apiKey, probe.toString(), new OpenAIClient.Callback() {
                @Override
                public void onSuccess(int status, String body, String requestId) {
                    boolean valid = status != 401 && status != 403;
                    onDone.done(valid, status == 200);
                }

                @Override
                public void onError(String code) {
                    onDone.done(false, false);
                }
            });
        } catch (Exception e) {
            onDone.done(false, false);
        }
    }
}
